pkg/resource/deploy/step_generator.go

```diff
@@ -333,172 +334,225 @@ func (sg *stepGenerator) GenerateSteps(event RegisterResourceEvent) ([]Step, res
 	//  It must be an update or a replace. Which operation we do depends on the the specific change made to the
 	//  resource's properties:
+	//
+	//  - if the user has requested that only specific resources be updated, and this resource is
+	//    not in that set, do no 'Diff' and just treat the resource as 'same' (i.e. unchanged).
+	//
 	//  - If the resource's provider reference changed, the resource must be replaced. This behavior is founded upon
 	//    the assumption that providers are recreated iff their configuration changed in such a way that they are no
 	//    longer able to manage existing resources.
+	//
 	//  - Otherwise, we invoke the resource's provider's `Diff` method. If this method indicates that the resource must
 	//    be replaced, we do so. If it does not, we update the resource in place.
 	if hasOld {
 		contract.Assert(old != nil)
-		diff, err := sg.diff(urn, old, new, oldInputs, oldOutputs, inputs, prov, allowUnknowns, goal.IgnoreChanges)
-		if err != nil {
-			// If the plugin indicated that the diff is unavailable, assume that the resource will be updated and
-			// report the message contained in the error.
-			if _, ok := err.(plugin.DiffUnavailableError); ok {
-				diff = plugin.DiffResult{Changes: plugin.DiffSome}
-				sg.plan.ctx.Diag.Warningf(diag.RawMessage(urn, err.Error()))
-			} else {
-				return nil, result.FromError(err)
+
+		// If the user requested only specific resources to update, and this resource was not in
+		// that set, then do nothin but create a SameStep for it.
+		if !shouldUpdate(updateTargetsOpt, urn) {
+			logging.V(7).Infof(
+				"Planner decided not to update '%v' due to not being in target group (same) (inputs=%v)", urn, new.Inputs)
+		} else {
+			updateSteps, res := sg.generateStepsFromDiff(
+				event, urn, old, new, oldInputs, oldOutputs, inputs, prov, goal)
+
+			if res != nil {
+				return nil, res
 			}
+
+			if len(updateSteps) > 0 {
+				// 'Diff' produced update steps.  We're done at this point.
+				return updateSteps, nil
+			}
+
+			// Diff didn't produce any steps for this resource.  Fall through and indicate that it
+			// is same/unchanged.
+			logging.V(7).Infof("Planner decided not to update '%v' after diff (same) (inputs=%v)", urn, new.Inputs)
 		}
 
-		// Ensure that we received a sensible response.
-		if diff.Changes != plugin.DiffNone && diff.Changes != plugin.DiffSome {
-			return nil, result.Errorf(
-				"unrecognized diff state for %s: %d", urn, diff.Changes)
+		// No need to update anything, the properties didn't change.
+		sg.sames[urn] = true
+		return []Step{NewSameStep(sg.plan, event, old, new)}, nil
+	}
+
+	// Case 4: Not Case 1, 2, or 3
+	//  If a resource isn't being recreated and it's not being updated or replaced,
+	//  it's just being created.
+	sg.creates[urn] = true
+	logging.V(7).Infof("Planner decided to create '%v' (inputs=%v)", urn, new.Inputs)
+	return []Step{NewCreateStep(sg.plan, event, new)}, nil
```

Comment:
What are your thoughts here?  We could proceed with teh 'create' (which will still show up in the preview).  Or we can error/warn here stating that there was an attempt to do a targetted update, but a resource was created outside of that set.

I personally think the former is fine.  it's akin to how targetted deletes might end up doing more work in some cases, but we let the user know.  I think targeted updates are intended to try to say that *updates* shouldn't happen to other resources.  I would also be fine with warnings saying that other non-update ops are happening to resources if we think the preview won't make it clear waht's happening.  But i personally don't think it's needed.

Reply:
What does Terraform do in this case?

Reply:
Will find out :)

Reply:
Sorry, got sidetracked.  Terraform *only* updates the requested resources.  Creates/deletes are not performed.

Reply:
Does the `terraform apply` still succeed?

Reply:
So here's what ends up happening.  

If you ask to target-update a resource that depends on a value from a created-resource, then tf will allow that created-resource to actually create.  However, if you don't end up depending on the created-resource, then it will not create it.

Not sure if we have the facility to do the same in our data-flow system.   For now, i've just added a block so that we don't proceed if we need to create a resource.  This feels non-onerous to me as a starting point.

If we did want to take the tf approach, I could def use some advice on how best i would implement that.  It feels like we would need some sort of "pending creates" list that we would put the create-steps in and then only execute if we were executing an update for something that depends on it.  It feels liek this is doable, but a lot of extra work to add support for.

Do you think we need it @pgavlin ?  Or would it be ok to start conservatively and then add support for that in the future if people ask for it?

Reply:
After speaking with some users, I do not believe that we need to take the TF approach. It seems easier to explain and understand the behavior you've implemented.

Reply:
(cc @lukehoban if he has anything to add)

Reply:
Thanks!  At the very least, i think this is a totally reasonable starting point.  We can def expand in the future if there is real demand for these more complex cases.  But 'm happy to have a reasonable set of behaviors that we can describe for destroy/refresh/update that likely solve hte 99% set of cases needed by the people upvoting this bug :)

Reply:
This plan makes sense to me - and seems conservative so we leave open the option to be more liberal in the future as we hear new use cases.

Reply:
Actually - thinking more about this - I'm a little more worried.

This means that if the program has diverged from the deployment by adding new resources, the `--target` can no longer be used to update the deployment?  I believe some variant of that is a meaningful % of the scenario for wanting to use this feature - right?

We should probably still land this as is - but I suspect without the ability to handle cases where unrelated resources have been added - the value of this will be reduced meaningfully.

Reply:
> the value of this will be reduced meaningfully.

I'm not sure how to make that call.  it comes down to how often we think: 1. someone will want to add a resource, 2. then they'll want to just update other existing resources.

Why can't they just update their existing resources first, then add the new resources?  It doesn't seem particularly onerous to have this restriction (at least as a starting point).

Note that we're not saying that this scenario can't/won't ever be supported.  Just that it doesn't seem as pressin to support from teh beginning, so it's easier to just hit more mainlike scenarios first at low cost and low complexity.